[#92] Support local development options.

dsummersl opened the pull request:
This commit supports the ability to specify different settings without
modifying files that are under version control. Changes:

 * By default the hsctl will use paths and settings specific to the
   current working directory and the current user. This allows the hctl
   command to work within VirtualBox, and outside it (if you have docker
   and gnu sed installed)

 * Introduces support for a .env file (see the README) to override
   various options (listed further on)

 * The ability to change which hydroshare-config.yaml is used by
   specifying a different path in the .env file.

 * The ability to use a different django settings file by specifying it
   in the .env file.

 * Creates a new dev_settings.example.py template that local developers
   can use to override/customize their environments in development.

 * Many settings.py settings that one would want to override in
   production or local development have been made configurable via
   environmental variables (but ideally one would do this via the .env
   file or their own custom settings.py file)

dsummersl committed:
[#92] Support local development options.

This commit supports the ability to specify different settings without
modifying files that are under version control. Changes:

 * By default the hsctl will use paths and settings specific to the
   current working directory and the current user. This allows the hctl
   command to work within VirtualBox, and outside it (if you have docker
   and gnu sed installed)

 * Introduces support for a .env file (see the README) to override
   various options (listed further on)

 * The ability to change which hydroshare-config.yaml is used by
   specifying a different path in the .env file.

 * The ability to use a different django settings file by specifying it
   in the .env file.

 * Creates a new dev_settings.example.py template that local developers
   can use to override/customize their environments in development.

 * Many settings.py settings that one would want to override in
   production or local development have been made configurable via
   environmental variables (but ideally one would do this via the .env
   file or their own custom settings.py file)

cbcunc commented:
I'm getting the following when I rebuild this branch from a clean clone and wiping out docker images and containers on a VirtBox instance:
```
hydro@hydro-VirtualBox:~/hydroshare$ docker logs hydroshare
groupmod: invalid group ID 'storage-hydro'
Unhandled exception in thread started by <function wrapper at 0x7f5ca34d45f0>
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/django/utils/autoreload.py", line 229, in wrapper
    fn(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/django/core/management/commands/runserver.py", line 107, in inner_run
    autoreload.raise_last_exception()
  File "/usr/local/lib/python2.7/site-packages/django/utils/autoreload.py", line 252, in raise_last_exception
    six.reraise(*_exception)
  File "/usr/local/lib/python2.7/site-packages/django/utils/autoreload.py", line 229, in wrapper
    fn(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/django/__init__.py", line 18, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/usr/local/lib/python2.7/site-packages/django/apps/registry.py", line 85, in populate
    app_config = AppConfig.create(entry)
  File "/usr/local/lib/python2.7/site-packages/django/apps/config.py", line 86, in create
    module = import_module(entry)
  File "/usr/local/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/usr/local/lib/python2.7/site-packages/mezzanine/conf/__init__.py", line 260, in <module>
    import_module("%s.defaults" % app)
  File "/usr/local/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/hydroshare/theme/defaults.py", line 1
    ../hydroshare/defaults.py
    ^
SyntaxError: invalid syntax
usermod: no changes
groupmod: invalid group ID 'storage-hydro'
Unhandled exception in thread started by <function wrapper at 0x7f4c60f625f0>
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/django/utils/autoreload.py", line 229, in wrapper
    fn(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/django/core/management/commands/runserver.py", line 107, in inner_run
    autoreload.raise_last_exception()
  File "/usr/local/lib/python2.7/site-packages/django/utils/autoreload.py", line 252, in raise_last_exception
    six.reraise(*_exception)
  File "/usr/local/lib/python2.7/site-packages/django/utils/autoreload.py", line 229, in wrapper
    fn(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/django/__init__.py", line 18, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/usr/local/lib/python2.7/site-packages/django/apps/registry.py", line 85, in populate
    app_config = AppConfig.create(entry)
  File "/usr/local/lib/python2.7/site-packages/django/apps/config.py", line 86, in create
    module = import_module(entry)
  File "/usr/local/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/usr/local/lib/python2.7/site-packages/mezzanine/conf/__init__.py", line 260, in <module>
    import_module("%s.defaults" % app)
  File "/usr/local/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/hydroshare/theme/defaults.py", line 1
    ../hydroshare/defaults.py
    ^
SyntaxError: invalid syntax
hydro@hydro-VirtualBox:~/hydroshare$ 
```

dsummersl committed:
[#92] Provide a default group id convention.

The virtualbox instance doesn't actually have a $GID field, so I guess
it isn't a standard. But it is fairly standard to have a group with the
same id as the user. Either way it can be overrid by a value in .env

dsummersl commented:
@cbcunc I wasn't able to reproduce this myself, but I did see that the group id wasn't getting set properly in the VM. I've pushed up a fix for that...and this is what I did to test it, which worked for me...

```shell
# start up the virtualbox (reboot if not a fresh startup

# open terminal and typed
cd hydroshare

# verify that the branch is 92-add-support-for-dev-settings
git branch

# remove any previously built docker instances:
./scripts/cleandocker

# start up hydroshare:
./hsctl rebuild --db

docker logs -f hydroshare
```

cbcunc reviewed (approved):
Tested on VirtBox OK. +1
